types/+MPCT/code_MPCT_ADMM_semiband_C.c

```diff
@@ -0,0 +1,930 @@
+
+    // Constant variables
+    $INSERT_CONSTANTS$
+
+    // Obtain variables in scaled units
+    #if in_engineering == 1
+    for(unsigned int i = 0; i < nn_; i++){
+        x0[i] = scaling_x[i]*( x0_in[i] - OpPoint_x[i] );
+        xr[i] = scaling_x[i]*( xr_in[i] - OpPoint_x[i] );
+    }
+    for(unsigned int i = 0; i < mm_; i++){
+        ur[i] = scaling_u[i]*( ur_in[i] - OpPoint_u[i] );
+    }
+    #endif
+    #if in_engineering == 0
+    for(unsigned int i = 0; i < nn_; i++){
+        x0[i] = x0_in[i];
+        xr[i] = xr_in[i];
+    }
+    for(unsigned int i = 0; i < mm_; i++){
+        ur[i] = ur_in[i];
+    }
+    #endif
+
+    // Compute b
+
+    for(unsigned int i = 0 ; i< nn_ ; i++){
+        b[i] = x0[i];
+    }
+
+
+    // Compute q
+
+    for(unsigned int i = 0 ; i < nn_ ; i++){
+        
+        for(unsigned int j = 0 ; j < nn_ ; j++){
+
+            q[NN_*nm_ + i] -= T[i][j] * xr[j];
+
+        }
+
+    }
+
+    for(unsigned int i = 0 ; i < mm_ ; i++){
+        
+        for(unsigned int j = 0 ; j < mm_ ; j++){
+
+            q[NN_*nm_ + nn_ + i] -= S[i][j] * ur[j];
+
+        }
+
+    }
+
+    // Measure time
+    #if MEASURE_TIME == 1
+
+    #if WIN32
+    QueryPerformanceCounter(&post_update); // Get time after the update    
+    t_update_time = 1000000000ULL * (post_update.QuadPart - start.QuadPart) / frequency.QuadPart;
+    sol->update_time = t_update_time/(double)1e+9;
+    #else // If Linux
+    clock_gettime(CLOCK_MONOTONIC_RAW, &post_update);
+    sol->update_time = (double) ( (post_update.tv_sec - start.tv_sec) * 1000.0 ) + (double) ( (post_update.tv_nsec - start.tv_nsec) / 1000000.0 );
+    #endif
+
+    #endif
+
+    // Algorithm
+    while(done == 0){
+
+        k += 1;
+
+        // Save the value of v
+        memcpy(v_old, v, sizeof(double)*(NN_+1)*nm_);
+        
+        // Reset acumulator variables
+        memset(z1_ac, 0, sizeof(double)*(NN_+1)*nm_);
+        memset(z3_ac, 0, sizeof(double)*(NN_+1)*nm_);
+        memset(z1_b, 0, sizeof(double)*(NN_+2)*nn_);
+        memset(z3_b, 0, sizeof(double)*(NN_+2)*nn_);
+        memset(z2, 0, sizeof(double)*2*nm_); 
+
+        //********** Equality-constrained QP solve **********//
+        // This problem updates z
+        for(unsigned int i = 0; i < (NN_+1)*nm_ ; i++){
+
+            #ifdef SCALAR_RHO
+            p[i] = q[i] + lambda[i] - rho * v[i];
+            #else
+            p[i] = q[i] + lambda[i] - rho[i] * v[i];
+            #endif
+
+        }
+
+
+        /****** Compute xi from eq. (9a) using Alg. 2 from the article ******/
+
+        solve_banded_diag_sys(Q_rho_i, R_rho_i, S_rho_i, T_rho_i, z1_ac, p); // Obtains z1_a
+
+        // z2_a = M_hat * z1_a computed sparsely
+        for (unsigned int i = 0 ; i < nn_ ; i++){
+
+            for (unsigned int l = 0 ; l < NN_+1 ; l++){
+
+                for (unsigned int j = 0 ; j < nn_ ; j++){
+
+                    if (l == NN_){
```

Comment:
This is very inefficient, because you are doing the `if(l == NN_)` check every single iteration. However, it is only satisfied once...  It would be much better to divide this into its two cases (one for `l == NN_` and one for all the rest of the values of `l`).